### PR TITLE
Optimization Function

### PR DESCRIPTION
**Pull request by DeepakSingh98**
#14 Added optimization function which can take inputs in any order and optimize only selected ones with a chosen loss function. Will add a footnote explaining the algorithm once I learn how population generation works.

### CODE REVIEW

File: py/optimize.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  YORKSHIRE = input_dict['YORKSHIRE']
+  EAST_MIDLANDS = input_dict['EAST_MIDLANDS']
+  WEST_MIDLANDS = input_dict['WEST_MIDLANDS']
+  EAST_OF_ENGLAND = input_dict['EAST_OF_ENGLAND']
+  LONDON = input_dict['LONDON']
+  SOUTH_EAST = input_dict['SOUTH_EAST']
+  SOUTH_WEST = input_dict['SOUTH_WEST']
+  WALES = input_dict['WALES']
+  SCOTLAND = input_dict['SCOTLAND']
+  NORTHER_IRELAND = input_dict['NORTHERN_IRELAND']
```
Comment (MaxGhenis) [inline py/optimize.py]:
```suggestion
  NORTHERN_IRELAND = input_dict['NORTHERN_IRELAND']

```
and below

File: py/optimize.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  # Declare bounds for each amount
+  bounds = [senior, child, dis_1, dis_2, dis_3, NORTH_EAST, NORTH_WEST, YORKSHIRE, EAST_MIDLANDS,
+            WEST_MIDLANDS, EAST_OF_ENGLAND, LONDON, SOUTH_EAST, SOUTH_WEST, WALES, SCOTLAND, NORTHER_IRELAND]
+  
+  # Take the average value of each tuple to create array
```
Comment (MaxGhenis) [inline py/optimize.py]:
```suggestion
  # Take the average value of each tuple to create array of starting values

```

File: py/optimize.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+from scipy.optimize import differential_evolution
```
Comment (MaxGhenis) [inline py/optimize.py]: think you need to add
```
from loss_functions import loss_metrics
```

File: py/optimize.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                 for each category. If min == max, the amount is fixed.
+  - loss_metric = the type of loss metric to be used in optimization.
+
+  Returns:
+  - prints the loss and corresponding UBI amount per category solution set for each function evaluation.
+  - return an OptimizeResult with the optimal solution.
+  '''
+  
+  # Declare categories 
+  categories = ['senior', 'adult', 'child', 'dis_1', 'dis_2', 'dis_3','NORTH_EAST', 'NORTH_WEST',
```
Comment (MaxGhenis) [inline py/optimize.py]:
```suggestion
  CATEGORIES = ['senior', 'adult', 'child', 'dis_1', 'dis_2', 'dis_3','NORTH_EAST', 'NORTH_WEST', 

```
(as a constant)

File: py/optimize.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  EAST_OF_ENGLAND = input_dict['EAST_OF_ENGLAND']
+  LONDON = input_dict['LONDON']
+  SOUTH_EAST = input_dict['SOUTH_EAST']
+  SOUTH_WEST = input_dict['SOUTH_WEST']
+  WALES = input_dict['WALES']
+  SCOTLAND = input_dict['SCOTLAND']
+  NORTHER_IRELAND = input_dict['NORTHERN_IRELAND']
+
+  # Declare bounds for each amount
+  bounds = [senior, child, dis_1, dis_2, dis_3, NORTH_EAST, NORTH_WEST, YORKSHIRE, EAST_MIDLANDS,
```
Comment (MaxGhenis) [inline py/optimize.py]: could this instead be this? could skip lines 22-38 if so
```
bounds = [input_dict[i] for i in CATEGORIES]
```

**Review by MaxGhenis — COMMENTED** [review]
Thanks @DeepakSingh98. Could you run `black` on the file? @nikhilwoodruff and I use it as our vscode formatter, which works well and should get rid of these pep8 errors.

If you've tested this in a notebook could you also upload that? If not it could be a separate PR from another of us, but would be good to know it runs before merging.

File: py/optimize.py (hunk 5 above)
Comment (DeepakSingh98) [reply py/optimize.py]: The bounds have to be in the same order as x, and since there's no guarantee the order of keys in input_dict will match that of CATEGORIES, I thought I needed lines 22-38 to reorder the variables into the specific order needed for x. Even if you make CATEGORIES = list(input_dict.keys()), I think you would need to reorder the keys to match the order of x before you can assign the variables using bounds = [input_dict[i] for i in CATEGORIES].

Comment (MaxGhenis) [reply py/optimize.py]: I don't follow, this seems to work:
```
In [1]: input_dict = {'adult': (0, 200), 'child': (0, 150), 'senior': (0, 250)}

In [2]: CATEGORIES = ['senior', 'adult', 'child']

In [3]: bounds = [input_dict[i] for i in CATEGORIES]

In [4]: bounds
Out[4]: [(0, 250), (0, 200), (0, 150)]
```

Comment (DeepakSingh98) [reply py/optimize.py]: Sorry, I misunderstood - I'll add this in now.

Comment (DeepakSingh98) [reply py/optimize.py:43]: Sorry, I misunderstood - I'll add this in now.

**Comment by DeepakSingh98** [comment]
> Thanks @DeepakSingh98. Could you run `black` on the file? @nikhilwoodruff and I use it as our vscode formatter, which works well and should get rid of these pep8 errors.
> 
> If you've tested this in a notebook could you also upload that? If not it could be a separate PR from another of us, but would be good to know it runs before merging.

I'm putting together an example notebook (just working on generating the datasets with frs right now), and I'll upload that when its done.

**Comment by DeepakSingh98** [comment]
> > Thanks @DeepakSingh98. Could you run `black` on the file? @nikhilwoodruff and I use it as our vscode formatter, which works well and should get rid of these pep8 errors.
> > If you've tested this in a notebook could you also upload that? If not it could be a separate PR from another of us, but would be good to know it runs before merging.
> 
> I'm putting together an example notebook (just working on generating the datasets with frs right now), and I'll upload that when its done.

Uploaded the notebook now - optimization runs (very fast), but the amount for seniors isn't obeying the bounds that were passed.

**Comment by DeepakSingh98** [comment]
The adult amount is being calculated on lines 223-225 of calc_ubi.py, but is never returned through to update x so we can correctly print the output_dict on lines 35-37 of optimize.py. I tried returning the adult amount through set_ubi and then loss_metrics but this causes an error since loss_metrics has to return a single value for the optimization to work.

Ideally, x[1] would be updated to equal the adult amount after it is calculated so the output_dict prints correctly. Any ideas on how best to extract the adult amount from set_ubi?

**Comment by nikhilwoodruff** [comment]
```diff
diff --git a/py/loss_functions.py b/py/loss_functions.py
index 094673b..2380586 100644
--- a/py/loss_functions.py
+++ b/py/loss_functions.py
@@ -3,7 +3,7 @@ import pandas as pd
 import microdf as mdf
 
 # File in repo.
-from calc_ubi import get_data, set_ubi
+from uk.py.calc_ubi import get_data, set_ubi
 
 
 baseline_df, reform_base_df, budget = get_data()

```

File: py/loss_functions.py [hunk 6]
```diff
@@ -29,8 +29,9 @@ def loss_metrics(x: list) -> pd.Series:
     :rtype: pd.Series
     """
     # Extract parameters and generate reform DataFrame.
-    senior, child, dis_1, dis_2, dis_3 = x[:5]
-    regions = np.array(x[5:])
+    senior = x[0]
+    child, dis_1, dis_2, dis_3 = x[2:6]
+    regions = np.array(x[6:])
     reform_df = set_ubi(
         reform_base_df, budget, senior, child, dis_1, dis_2, dis_3, regions
```
Comment (MaxGhenis) [inline py/loss_functions.py:37]:
```suggestion
        reform_base_df, budget, senior, child, dis_1, dis_2, dis_3, regions, verbose=True
```
(if you want to print out the adult amount)

**Review by MaxGhenis — COMMENTED** [review]
.

**Comment by MaxGhenis** [comment]
Or are you looking for a function to extract the adult UBI, not just print it? @nikhilwoodruff is there a clean way to pull out the adult amount calculation from `set_ubi` in a way that can be used here?

**Comment by nikhilwoodruff** [comment]
```diff
diff --git a/py/loss_functions.py b/py/loss_functions.py
index 9e790e2..e59e753 100644
--- a/py/loss_functions.py
+++ b/py/loss_functions.py
@@ -3,7 +3,7 @@ import pandas as pd
 import microdf as mdf
 
 # File in repo.
-from calc_ubi import get_data, set_ubi
+from uk.py.calc_ubi import get_data, set_ubi
 
 
 baseline_df, reform_base_df, budget = get_data()

```

File: py/loss_functions.py (hunk 6 above)
Comment (DeepakSingh98) [reply py/loss_functions.py:37]: > Or are you looking for a function to extract the adult UBI, not just print it? @nikhilwoodruff is there a clean way to pull out the adult amount calculation from `set_ubi` in a way that can be used here?

Ideally I'd like to extract the adult amount from set_ubi so that I can set x[1] = adult_amount before creating and printing the output_dict.

Comment (MaxGhenis) [reply py/loss_functions.py:37]: > Uploaded the notebook now - optimization runs (very fast), but the amount for seniors isn't obeying the bounds that were passed.

Weird. I can't find anything about this issue. If we can reproduce could be worth filing an issue in the scipy repo. In the meantime, could you make the function value `np.inf` (or some big constant) for scenarios that violate the constraints? Seems like this would be necessary to respect the adult amount bounds anyway (though that doesn't seem to be an issue).

Could you prioritize this over getting the adult amount in the print message?

Comment (DeepakSingh98) [reply py/loss_functions.py:37]: I've managed to get the optimization working and I've uploaded a notebook showing two different optimizations

The adult amount was being calculated but then not being used in the actual optimization, which was causing problems (in addition to an error with ordering the output_dict), so I created a function in calc_ubi.py to extract the adult amount directly and then feed it into the optimization. Currently the optimization works so long as you don't fix any variables (see the third example in the notebook where I try to fix the regional supplements). You can see from the output that it progressed somewhat (with the amounts obeying the bounds) before exiting with the error.

Calling get_data() in optimize.py and the extract() function are both areas which could probably be simplified by moving them to just optimize.py instead of having them in loss_functions.py too, but the optimization still runs fast.

**Comment by MaxGhenis** [comment]
Cool, thanks @DeepakSingh98. Too bad about the fixed bounds breaking the algorithm, guess we'll need another approach.

Does `adult` needs to be part of `x`? Seems like the algorithm is just discarding it anyway right? Adding the extra dimension could be slowing down the algorithm, and it seems like it means the ultimate `OptimizeResult` `x` produces an incorrect `adult` value (needs to be recalculated based on the rest of `x`).

**Comment by MaxGhenis** [comment]
To see if I interpret the [notebook](https://nbviewer.jupyter.org/github/UBICenter/uk/blob/1a6b927da13bed44adf7b459c2ebc36328b055f0/Optimization_with_Differential_Evolution_V3.ipynb) correctly though, seems like minimizing % losers produces an optimal of 0.17% losers (seems really low to me...) with UBI structure:
* 177 for seniors
* ? for adults (shows 117 but I don't think that'd be correct)
* 87 for kids
* 15/47/72 for 3 disability levels
* plus geo supplements

Minimizing % loss produces an average percent loss of 1.4%, with optimal levels of:
* 57 for seniors
* ? for adults
* 110 for kids
* 9/20/45 for the 3 levels of disability
* plus geo supplements

Not necessary for now since we're going to go with the housing supplement levels, but I also think it'd be good to use a base geography for calculating UBI levels. It's doing more searching than necessary with all 12 moving together.

File: py/optimize.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  bounds = [input_dict[i] for i in CATEGORIES]
+
+  # Take the average value of each tuple to create array of starting values
+  x = [((i[0] + i[1])/2) for i in bounds]
+
+  def loss_func(x, args=(loss_metric)):
+    loss_metric = args
+
+    # Calculate loss with current solution.
+    loss = loss_metrics(x)[loss_metric]
```
Comment (nikhilwoodruff) [inline py/optimize.py]: @DeepakSingh98 think this might be a bug: we're calculating the loss, and then retrospectively changing the adult amount. Could the optimization function be getting really low losses by giving high amounts to adults, getting low losses, and then constraining the adult amount?

File: py/calc_ubi.py [hunk 8]
```diff
@@ -235,9 +218,37 @@ def set_ubi(
...
+        dis_1 (float): Disabled (Equality Act+) supplement per week
+        dis_2 (float): Enhanced disabled supplement per week
+        dis_3 (float): Severely disabled supplement per week
+        regions (ndarray): Numpy float array of 12 UK regional supplements per week
+        verbose (bool, optional): Whether to print the calibrated adult UBI amount. Defaults to False.
+
+    Returns:
+        DataFrame: Reform household-level DataFrame
+    """
+    basic_income, adult_amount = get_adult_amount(base_df, budget, senior, child, dis_1, dis_2, dis_3, regions, pass_income=True)
```
Comment (nikhilwoodruff) [inline py/calc_ubi.py]: Minor suggestions: maybe pass on the verbosity argument?
```suggestion
    basic_income, adult_amount = get_adult_amount(base_df, budget, senior, child, dis_1, dis_2, dis_3, regions, pass_income=True, verbose=verbose)
```

File: py/optimize.py (hunk 7 above)
Comment (nikhilwoodruff) [reply py/optimize.py]: Actually, never mind - the loss function itself does do the calibration as well.

**Comment by MaxGhenis** [comment]
@nikhilwoodruff and I found some errors (largely or entirely in my code I think), he'll add some of those changes.

File: py/calc_ubi.py [hunk 9]
```diff
@@ -183,11 +183,11 @@ def get_data(path=None):
     else:
         person, benunit, household = frs.load()
     baseline = PopulationSim(frs_data=(person, benunit, household))
-    baseline_df = calc2df(baseline, BASELINE_COLS)
+    baseline_df = calc2df(baseline, BASELINE_COLS, map_to="household")
     FRS_DATA = (person, benunit, household)
     reform_no_ubi = ubi_reform(0, 0, 0, 0, 0, 0, np.array([0] * 12))
     reform_no_ubi_sim = PopulationSim(reform_no_ubi, frs_data=FRS_DATA)
-    reform_base_df = calc2df(reform_no_ubi_sim, BASELINE_COLS)
+    reform_base_df = calc2df(reform_no_ubi_sim, BASELINE_COLS, map_to="household")
```
Comment (nikhilwoodruff) [inline py/calc_ubi.py]: Mapping to household to avoid over-counting (total population was 200m before)

File: py/optimize.py [hunk 10]
```diff
@@ -26,14 +24,17 @@ def optimize(input_dict, loss_metric):
...
+
   # Take the average value of each tuple to create array of starting values
   x = [((i[0] + i[1])/2) for i in bounds]
 
   def loss_func(x, args=(loss_metric)):
     loss_metric = args
 
     # Calculate loss with current solution.
-    loss = loss_metrics(x)[loss_metric]
+    loss = loss_metrics(x, baseline_df, reform_base_df, budget)[loss_metric]
```
Comment (nikhilwoodruff) [inline py/optimize.py]: Pass in budget variables rather than globals just for peace of mind

File: py/loss_functions.py [hunk 11]
```diff
@@ -57,22 +54,12 @@ def loss_metrics(x: list) -> pd.Series:
     total_pct_loss = np.sum(weight[valid_pct_loss] * pct_loss[valid_pct_loss])
     mean_pct_loss = total_pct_loss / total_pop
     # Calculate average percent loss with double weight for PWD.
-    pwd2_weight = weight * np.where(baseline_df.is_disabled, 2, 1)
+    pwd2_weight = baseline_df.household_weight * (baseline_df.is_disabled + baseline_df.people_in_household)
```
Comment (nikhilwoodruff) [inline py/loss_functions.py]: I think this fixes the disabled counting problem...

File: py/calc_ubi.py [hunk 12]
```diff
@@ -247,7 +247,7 @@ def set_ubi(
     """
     basic_income, adult_amount = get_adult_amount(base_df, budget, senior, child, dis_1, dis_2, dis_3, regions, pass_income=True)
     basic_income += base_df["is_WA_adult"] * adult_amount
-    reform_df = base_df
+    reform_df = base_df.copy(deep=True)
```
Comment (nikhilwoodruff) [inline py/calc_ubi.py:314]: Maybe just for safety...

**Review by nikhilwoodruff — COMMENTED** [review]
Just summarising the changes

**Comment by MaxGhenis** [comment]
btw don't worry about the failures, they're due to https://github.com/executablebooks/jupyter-book/issues/1137 and will be fixed with jupyter-book issues its next release (there's a band-aid available but since we're not using the jb right now not really needed).

**Comment by MaxGhenis** [comment]
Following up on our chat, couple next steps here:
* Add `**kwargs` to `optimize` which would be passed to `differential_evolution`, e.g. so we can try different `tol` or `maxiter` values
* Use the above to ensure it runs long enough such that the loss always falls when adding degrees of freedom
* Expand bounds on geographic supplements, since some are hitting the max of 30
* [Nice to have] Remove `adult` from `x`
* [Nice to have] Print out the actual adult amount, or include an example of how to use the `optimize` result to get the adult amount (seems like `verbose=True` isn't happening)

**Comment by DeepakSingh98** [comment]
**Summarizing recent changes:**

optimize.py
- added print_output argument (toggles printing of function evaluations) 
- added **kwargs which are passed to differential_evolution
- expanded geographic supplement bounds to (-30, 45)
- removed adult from x
- inserted correct adult amount into the optimal solution set in the OptimizeResult that is returned by optimize()

loss_functions.py
- changed slicing in extract() to reflect the removal of adult from x

calc_ubi.py
- no longer passing verbose since the adult amount is inserted in the optimal x in the OptimizeResult object

Reform_Level_Optimization_V2.ipynb
- notebook (only optimal solutions printed) showing optimization

**Next Steps**
- potentially add in the correct adult amount for each function evaluation output_dict (currently avoiding this to speed up optimization considerably)
- experiment with different tol values to see if loss decreases further (currently loss does decrease with additional degrees of freedom, but optimization exits successfully after iterations 8, 9, and 13 for reforms 1, 2, and 3 respectively, with a set maxiter=20)

**Comment by DeepakSingh98** [comment]
Additional changes to optimize.py:
- "print_output" changed to "verbose" (had avoided previously to avoid confusion but since there's no need to use the verbose argument for the adult amount this should be fine)
- when verbose=True, the printed output_dict now contains the correct adult amount (takes much longer to run however)
- optimal solution dict is printed regardless of verbose for clarity

**Comment by DeepakSingh98** [comment]
Added loss_dict which shows the various losses by different metrics for the optimal x found through optimization for the chosen metric.

**Comment by MaxGhenis** [comment]
Merging this so we can move forward in this repo rather than Deepak's branch (this mega-PR madness has gone on too long!)
